### PR TITLE
core: refactor to use Shared trait

### PR DESCRIPTION
**Pull request by ry**
Introduces the Shared trait which abstracts access to the shared buffer. SharedSimple as a very basic implementation of this which is used in deno_core_http_bench

### CODE REVIEW

**Comment by afinch7** [comment]
Looks great. `record.x` is a lot more friendly(and potentially less error prone) than the old `get_record(i, off)` and `set_record(i, off, value)`, also allowing for alternative shared implementations is not a bad idea in case you come up with a way to make this faster in the future. Core looks like it's getting pretty close to ready to replace the old isolate implementation.

File: core/lib.rs [hunk 1]
```diff
@@ -6,71 +6,103 @@ extern crate libc;
...
+  result: Option<R>,
+}
+
+impl<R> PendingOp<R> {
+  /// Calls poll() on a PendingOp. Returns true if it's complete.
+  fn is_complete(self: &mut PendingOp<R>) -> bool {
+    // TODO the name of this method isn't great, because it doesn't indicate
+    // that the op is being polled.
+
+    // Do not call poll on futures we've already polled this turn.
```
Comment (piscisaureus) [inline core/lib.rs]: How come we are polling the same future twice on the same turn? 
(I'm worried that that doesn't scale well if there are many outstanding ops)

File: core/http_bench.rs [hunk 2]
```diff
@@ -53,10 +57,10 @@ fn main() {
 
   let args: Vec<String> = env::args().collect();
   if args.len() > 1 && args[1] == "--multi-thread" {
-    println!("multi-thread");
+    debug!("multi-thread");
     tokio::run(main_future);
   } else {
-    println!("single-thread");
```
Comment (ry) [inline core/http_bench.rs]: Keep println

File: core/lib.rs (hunk 1 above)
Comment (ry) [reply core/lib.rs]: Ops aren’t polled more than once. 

Imagine the situation where an accept is followed by a read. Once the accept has completed and before the read has been added, mio has not been notified to wake up. Failing to poll the newly added read op will result in incorrect behavior.

**Comment by ry** [comment]
closed in favor of #1888

**Pull request closed without merging**